### PR TITLE
Query: Fixes handling of pipeline execution on partition merge

### PR DESCRIPTION
**Pull request by ealsur**
# Pull Request Template

## Description

The `CrossPartitionRangePageAsyncEnumerator` when facing a merge, was adding the merge result as part of the list of Enumerators, instead of maintaining the current Range and draining it.

In this sample scenario, let's assume a 2 partition collection (partition A and B).

When the query pipeline starts, it generates a queue with 2 enumerators, one for A, one for B. Assume each partition has 10 documents.

Assume we fetch the first page which contains 10 documents from A, which returns a ContinuationToken like:

`{token: null, range {B}}`

Which normally would make the query continue with B and drain it.

If there is a merge and A and B merge into C (with C having all the 20 documents), the current logic adds a new enumerator for C in the queue copying the continuation from the parent (B), so the query would not continue on C (full range) with continuation `null` (all the documents).

The effect is that the query end ups returning 30 documents (10 from the first page, 20 from the second because it used C).

### What is the fix

In the case of a merge where the result is a single parent range, instead of creating a new Queue Enumerator for the merge result, we re-enqueue the current Range Enumerator. 

When this happens, the logic in the RequestInvokerHandler kicks in and applies the EPK filtering:

https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs#L244-L263

There was a test currently trying to cover this scenario but it was failing to catch the bug due to a bug in the test itself and how the InMemoryContainer behaved on a merge.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks

File: Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs [hunk 1]
```diff
@@ -184,11 +184,21 @@ public async ValueTask<bool> MoveNextAsync(ITrace trace)
                                 error: new Microsoft.Azure.Documents.Error { Code = "SDK_invariant_violated_4795CC37", Message = errorMessage });
                         }
 
-                        foreach (FeedRangeInternal childRange in childRanges)
+                        if (childRanges.Count == 1)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs:187]: Any possibility of a merge then a split? Just trying to think of any scenario where a merge might not result in a single child range.

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs:187]: the childRanges are calculated over the overlap between the system ranges and the desired range.

If there was a split and merge or merge and split it doesn't matter, what matters is, is the current range spanning 1 or more system ranges? If it spans 1, then keep using the current range, if it spans more than 1, then split it.

**Review by FabianMeiswinkel — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs [hunk 2]
```diff
@@ -84,6 +242,7 @@ private async Task<Exception> ShouldReturnFailure()
                         cancellationToken: default);
 
                     await this.DocumentContainer.MergeAsync(ranges[0], ranges[1], default);
+                    await this.DocumentContainer.RefreshProviderAsync(NoOpTrace.Singleton, default);
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs:245]: >await this.DocumentContainer.RefreshProviderAsync(NoOpTrace.Singleton, default); [](http://example.com/codeflow?start=20&length=80)

why is this required here ?

**Review by neildsh — APPROVED** [review]

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs (hunk 2 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs:245]: Found out that the InMemoryContainer maintains a cached map here https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/InMemoryContainer.cs#L208,

When MergeAsync is called, it updates the non-cached ranges https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/InMemoryContainer.cs#L1024 but never the cached ones, so any operation afterwards still uses the cached values which do not reflect the merge.

**Review by j82w — APPROVED** [review]
